questdb-rs/src/ingress/ndarr.rs

```diff
@@ -0,0 +1,754 @@
+    // `std::ptr::copy_nonoverlapping` on both arm(Macos) and x86(Linux) platform.
+    // This may because `copy_from_slice` benefits more from compiler.
+    if let Some(contiguous) = array.as_slice() {
+        let bytes = unsafe {
+            slice::from_raw_parts(contiguous.as_ptr() as *const u8, size_of_val(contiguous))
+        };
+
+        if bytes.len() != expect_size {
+            return Err(error::fmt!(
+                ArrayWriteToBufferError,
+                "Array write buffer length mismatch (actual: {}, expected: {})",
+                expect_size,
+                bytes.len()
+            ));
+        }
+
+        if buf.len() < bytes.len() {
+            return Err(error::fmt!(
+                ArrayWriteToBufferError,
+                "Buffer capacity {} < required {}",
+                buf.len(),
+                bytes.len()
+            ));
+        }
+
+        buf[..bytes.len()].copy_from_slice(bytes);
+        return Ok(());
+    }
+
+    // Fallback path: non-contiguous memory handling
+    // For non-contiguous memory layouts, direct raw pointer operations are preferred.
+    let elem_size = size_of::<T>();
+    let mut total_len = 0;
+    for (i, &element) in array.iter().enumerate() {
+        unsafe {
+            std::ptr::copy_nonoverlapping(
+                &element as *const T as *const u8,
+                buf.as_mut_ptr().add(i * elem_size),
+                elem_size,
+            )
+        }
+        total_len += elem_size;
+    }
+    if total_len != expect_size {
+        return Err(error::fmt!(
+            ArrayWriteToBufferError,
+            "Array write buffer length mismatch (actual: {}, expected: {})",
+            total_len,
+            expect_size
+        ));
+    }
+    Ok(())
+}
+
+#[cfg(feature = "benchmark")]
+pub(crate) fn write_array_data_use_raw_buffer<A: NdArrayView<T>, T>(buf: &mut [u8], array: &A)
+where
+    T: ArrayElement,
+{
+    // First optimization path: write contiguous memory directly
+    if let Some(contiguous) = array.as_slice() {
+        let byte_len = size_of_val(contiguous);
+        unsafe {
+            std::ptr::copy_nonoverlapping(
+                contiguous.as_ptr() as *const u8,
+                buf.as_mut_ptr(),
+                byte_len,
+            )
+        }
+    }
+
+    // Fallback path: non-contiguous memory handling
+    let elem_size = size_of::<T>();
+    for (i, &element) in array.iter().enumerate() {
+        unsafe {
+            std::ptr::copy_nonoverlapping(
+                &element as *const T as *const u8,
+                buf.as_mut_ptr().add(i * elem_size),
+                elem_size,
+            )
+        }
+    }
+}
+
+pub(crate) fn get_and_check_array_bytes_size<A: NdArrayView<T>, T>(
+    array: &A,
+) -> Result<usize, Error>
+where
+    T: ArrayElement,
+{
+    (0..array.ndim())
+        .try_fold(std::mem::size_of::<T>(), |acc, i| Ok(acc * array.dim(i)?))
+        .and_then(|p| match p <= MAX_ARRAY_BUFFER_SIZE {
```

Comment:
Match over bool is just an `if` statement.
I also think this code would probably be easier to read if it were iterative.

```rust
let mut size = std::mem::size_of::<T>();
for dim_index in 0..array.ndim() {
    size = size * array.dim(dim_index)?
    if size > MAX_ARRAY_BUFFER_SIZE {
        return Err(...);
    }
}
Ok(size)
```